Deploy scripts and readme

simonwheatley opened the pull request:
A deploy script which can be called from either Circle CI or Travis CI to create a `${BRANCH}-built` version, and push it to GitHub.

simonwheatley commented:
@rmccue I ended up using but then significantly departing from your scripts, please let me know if there's any issues with what we've ended up with.

joshbetz commented on ci/deploy.sh:
```diff
@@ -0,0 +1,100 @@
+fi
+
+echo "Syncing files... quietly"
+rsync -a "$SRC_DIR/" "$BUILD_DIR" --exclude-from "$SRC_DIR/ci/deploy-exclude.txt"
```
Should we `--cvs-exclude` in addition or instead of defining a list?

simonwheatley replied:
Addressed in 2bf1ae0d1b0c327256169e9f2d86b1a3132a3928

joshbetz commented:
I think we should always exclude cvs files rather than depending on a config that's defined in the repo, but otherwise looks good.

joshbetz commented on ci/deploy.sh:
```diff
@@ -0,0 +1,100 @@
+DEPLOY_SUFFIX="-built"
+
+BRANCH="${CIRCLE_BRANCH:-$TRAVIS_BRANCH}"
+SRC_DIR="${TRAVIS_BUILD_DIR:-$PWD}"
```
I'm assuming we'd have to manually change these on Circle CI?

simonwheatley replied:
I'm not sure what you mean. I've added a comment explaining the `CIRCLE_*` and `TRAVIS_*` environment variables are set by the CI provider. 

Looking into `$PWD`, it seems to be a somewhat reliable variable set by the shell. I could move to `PWD=$(pwd)` for more safety.

…or are you thinking of something else?

joshbetz replied:
Oh, I missed `CIRCLE_BRANCH` :|

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,70 @@
+Scripts designed to facilitate building your Javascript and CSS
+on either [Travis CI](https://travis-ci.com) or [Circle CI](https://circleci.com/).
+
+See our documentation here: (NEEDS LINK, Ed)
```
`(NEEDS LINK, Ed)` :)

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,70 @@
+  - 7.1
+
+script:
+  - find . -name \*.php -print0 | xargs -0 -n 1 -P 4 php -d display_errors=stderr -l > /dev/null
```
The linting isn't really relevant here?

simonwheatley replied:
I simplified the Travis example

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,70 @@
+  submodules: false
+
+php:
+  - 7.1
```
Why 7.1?

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,70 @@
+
+See our documentation here: (NEEDS LINK, Ed)
+
+Sample Circle CI config, put into `.circleci/config.yml`:
```
I think it would be helpful to point out the most important and relevant bits from the config files for repos that may already have have a config file set up (i.e. 1) branch ignore and 2) `steps`+`deploy` for Circle / `after_success` for Travis).

simonwheatley replied:
👍 I've added more commentary in the examples

mjangda commented on ci/deploy-exclude.txt:
```diff
@@ -0,0 +1,8 @@
+# Add files or patterns to exclude from the built branch here.
```
Should we name this `.deployignore` for consistency with `gitignore` and `svnignore`?

simonwheatley replied:
Sold. I've renamed it.

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,107 @@
+fi
+
+if [[ -d "$BUILD_DIR" ]]; then
+	echo "WARNING: ${BUILD_DIR} already exists. You may have accidentally cached this"
```
Should we prefix the BUILD_DIR with a timestamp to minimize potential failures due to caching?

simonwheatley replied:
I'm not sure about this, if we do this and the directory does get cached then we'll gradually fill the cache will directories with timestamps in the name. What do you think, @mjangda ?

mjangda replied:
I think the space issue is more a problem for the CI service (or the server running the script) to handle. We should optimize for reliable deploys (making sure we push out the right thing and try to limit failures, which can be annoying to recover from).

simonwheatley replied:
Sold. I've added a timestamp.

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,107 @@
+#    branches:
+#      ignore:
+#        - /^.*(?<!-built)$/
+if [[ "$BRANCH" == *${DEPLOY_SUFFIX} ]]; then
```
Minor: this check could be moved higher before we `cd` into `SRC_DIR`

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,107 @@
+COMMIT_USER_NAME="$( git log --format=%ce -n 1 $COMMIT_SHA )"
+COMMIT_USER_EMAIL="$( git log --format=%cn -n 1 $COMMIT_SHA )"
+GIT_USER="${DEPLOY_GIT_USER:-$COMMIT_USER_NAME}"
+GIT_EMAIL="${DEPLOY_GIT_EMAIL:-$COMMIT_USER_EMAIL}"
```
Where are `DEPLOY_GIT_USER` and `DEPLOY_GIT_USER` defined? If they are Travis / Circle-specific vars, we should document that inline.

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,107 @@
+GIT_USER="${DEPLOY_GIT_USER:-$COMMIT_USER_NAME}"
+GIT_EMAIL="${DEPLOY_GIT_EMAIL:-$COMMIT_USER_EMAIL}"
+
+git clone "$REPO_SSH_URL" "$BUILD_DIR"
```
A full clone is probably unnecessary (and slow for big repos)? Could likely get away with `init` + `remote add`.

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,107 @@
+fi
+
+# Ensure we're in the right dir
+cd "$BUILD_DIR"
```
We already did this earlier.

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,107 @@
+fi
+
+# Commit it.
+MESSAGE=$( printf 'Build changes from %s\n\n%s' "${COMMIT}" "${CIRCLE_BUILD_URL}" )
```
Minor: would be neat to support co-authors in a future version (https://help.github.com/articles/creating-a-commit-with-multiple-authors/)

simonwheatley commented:
Might actually be there. Would prefer to squash and merge this, when the time comes, to remove the messy history.

Would you review, please, @mjangda?

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,98 @@
+# VIP Go CI Scripts
+
+This directory is not deployed to your VIP site.
```
This comment seems out of place. Feels like we need an intro sentence instead first?

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,98 @@
+
+## Credits
+
+Inspiration (and some code) taken from [Human Made's VIP Go Builder](https://github.com/humanmade/vip-go-builder/), thank you!
```
❤️

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,98 @@
+
+### [Circle CI](https://circleci.com/)
+
+Sample Circle CI config, put into `.circleci/config.yml`:
```
We should call out the `steps` section (i.e. "configure your build steps"), since it's easy to gloss over. Otherwise, we'll have devs adding the config as-is and wondering why things don't just work :)

simonwheatley replied:
Addressed in https://github.com/Automattic/vip-go-skeleton/pull/20/commits/559ce1fc966f3e3160ee65f9c94238401fe75295

mjangda commented on ci/README.md:
```diff
@@ -0,0 +1,98 @@
+### [Travis CI](https://travis-ci.com)
+
+1. Add the config to your repository
+2. Create and add a deploy key to your repository settings on Travis
```
Same as with Circle, should call out the `before_script` section.

simonwheatley replied:
Addressed in https://github.com/Automattic/vip-go-skeleton/pull/20/commits/559ce1fc966f3e3160ee65f9c94238401fe75295

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,112 @@
+# ---------------
+
+if [[ -z "${BRANCH}" ]]; then
+	echo "No branch specified!"
```
For this, and the other fatal errors below, it would be nice to include some instructions or direction on what to do next. Clearly something went wrong, but what do I do about it? (Related: https://www.smashingmagazine.com/2011/05/getting-started-with-defensive-web-design/)

simonwheatley replied:
https://github.com/Automattic/vip-go-skeleton/pull/20/commits/4fbd17d7b66bf55344c2a45f96dd66a912b61c60

simonwheatley commented:
Addressed feedback, @mjangda. More feedback always welcome 🙂

My only remaining consideration is [how long the `README.md` is now](https://github.com/Automattic/vip-go-skeleton/blob/4fbd17d7b66bf55344c2a45f96dd66a912b61c60/ci/README.md), and whether the content should be moved largely into our documentation leaving barebones instructions and a link. I'm going to sleep on it.

rmccue commented:
(Just dropping a quick note to mention I've been following this, looks great! Glad the HM codebase provided useful inspiration! ❤️)

mjangda commented on ci/deploy.sh:
```diff
@@ -0,0 +1,122 @@
+# -------------------
+
+if ! command -v 'rsync'; then
+	sudo apt-get install -q -y rsync
```
We should run `apt-get update` to avoid issues with out-of-date dependencies/versions of `rsync` and other tools we're using. (This came up on a client site today.)

simonwheatley commented:
Should consider storing the main scripts in a separate repo, and including them via Curl or whatever during the CI execution, so we always get the most recent version with any bug fixes without having to update all client site repos.

simonwheatley commented:
Moved the scripts to https://github.com/Automattic/vip-go-build

simonwheatley commented:
This is no longer necessary. Documentation will be on our documentation site.
